src/vm/types.rs

```diff
@@ -70,143 +212,399 @@ impl TupleTypeSignature {
         }
         Ok(TupleData { type_signature: TupleTypeSignature { type_map: type_map },
                        data_map: data_map })
 
     }
 
-    pub fn get(&self, name: &str) -> InterpreterResult {
+    pub fn get(&self, name: &str) -> Result<Value> {
         if let Some(value) = self.data_map.get(name) {
             Ok(value.clone())
         } else {
             Err(Error::InvalidArguments(format!("No such field {:?} in tuple", name)))
         }
         
     }
+
+    pub fn size(&self) -> i128 {
+        self.type_signature.size()
+    }
+}
+
+impl fmt::Display for TupleData {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let mut first = true;
+        write!(f, "(")?;
+        for (name, value) in self.data_map.iter() {
+            if !first {
+                write!(f, ", ")?;
+            }
+            first = false;
+            write!(f, "{}: {}", name, value)?;
+        }
+        write!(f, ")")
+    }
 }
 
 impl TypeSignature {
-    pub fn new(atomic_type: AtomTypeIdentifier, dimension: u8) -> TypeSignature {
+    pub fn new_atom(atomic_type: AtomTypeIdentifier) -> TypeSignature {
         TypeSignature { atomic_type: atomic_type,
-                        dimension: dimension }
+                        list_dimensions: None }
+    }
+
+    pub fn new_list(atomic_type: AtomTypeIdentifier, max_len: i128, dimension: i128) -> Result<TypeSignature> {
+        if dimension == 0 {
+            Err(Error::InvalidTypeDescription)
+        } else if max_len > u32::max_value() as i128 || dimension > u8::max_value() as i128 {
+            Err(Error::ListTooLarge)
+        } else {
+            let list_dimensions = Some(ListTypeData { max_len: max_len as u32,
+                                                      dimension: dimension as u8 });
+            let type_sig = TypeSignature { atomic_type: atomic_type,
+                                           list_dimensions: list_dimensions };
+            if type_sig.size() > MAX_VALUE_SIZE {
+                Err(Error::ValueTooLarge)
+            } else {
+                Ok(type_sig)
+            }
+        }
+    }
+
+    fn new_atom_checked(atom_type: AtomTypeIdentifier) -> Result<TypeSignature> {
+        if atom_type.size() > MAX_VALUE_SIZE {
+            Err(Error::ValueTooLarge)
+        } else {
+            Ok(TypeSignature::new_atom(atom_type))
+        }
+    }
+
+    fn new_tuple(tuple_type_sig: TupleTypeSignature) -> Result<TypeSignature> {
+        TypeSignature::new_atom_checked(AtomTypeIdentifier::TupleType(tuple_type_sig))
+    }
+
+    fn new_buffer(buff_len: i128) -> Result<TypeSignature> {
+        if buff_len > u32::max_value() as i128 {
+            Err(Error::BufferTooLarge)
+        } else {
+            let atom_type = AtomTypeIdentifier::BufferType(buff_len as u32);
+            TypeSignature::new_atom_checked(atom_type)
+        }
+    }
+
+    pub fn get_empty_list_type() -> TypeSignature {
```

Comment:
Why is the `atomic_type` for an empty list an `IntType`?

Reply:
an empty list in this form has a maximum length of 0, meaning that it can't ever hold any values whatsoever. It should really be something like an "Any" type, though, so that if you had, say, a map value that required something of type `(list 1 bool)`, an empty list would fit, and it should fit in `(list 1 int)` as well.

Reply:
Okay, sounds good.